### PR TITLE
feat(notifications): Add url_format property to notifications

### PR DESCRIPTION
**Pull request by 97varun**
Add a `url_format` property to `BaseNotification` to support different URL formats across providers. This property should be set on the notification object by the provider's message builder which builds the notification messages.

### CODE REVIEW

File: src/sentry/notifications/notifications/base.py [hunk 1]
```diff
@@ -60,6 +60,20 @@ def reference(self) -> Model | None:
...
+        """
+        The URL format used when embedding links in text.
+        Slack notifications should set this to `<{url}|{text}>`.
+        Microsoft Teams notifications should set this to `[text](url)`.
+        URLs should be built using `self.url_format.format(text=some_text, url=some_url)`.
+        """
+        return self._url_format
+
+    @url_format.setter
+    def url_format(self, url_format: str) -> None:
```
[the remark this answers is not in the document]
Comment (97varun) [reply src/sentry/notifications/notifications/base.py:74]: Good idea, I would prefer throwing an error when it is not set in the getter, let me make that change.

[the remark this answers is not in the document]
Comment (97varun) [reply src/sentry/notifications/notifications/base.py:79]: It already threw an error if the url_format was not set, but I customized it a little - `AttributeError: 'url_format' not set on 'AssignedActivityNotification'. Please set the 'url_format' property from the message builder.`

File: src/sentry/integrations/slack/message_builder/notifications/base.py [hunk 2]
```diff
@@ -18,6 +18,7 @@ def __init__(
     ) -> None:
         super().__init__()
         self.notification = notification
+        self.notification.url_format = SLACK_URL_FORMAT
```
[the remark this answers is not in the document]
Comment (97varun) [reply src/sentry/integrations/slack/message_builder/notifications/base.py]: @NisanthanNanthakumar updated! Thanks to this, found another place where the notification properties have to be set.

https://github.com/getsentry/sentry/blob/c5cd407fd0ab2c7313ad72ddde09fb403637e48c/src/sentry/integrations/slack/notifications.py#L126-L128

**Review by NisanthanNanthakumar — APPROVED** [review]

File: src/sentry/notifications/notifications/base.py [hunk 3]
```diff
@@ -247,7 +272,25 @@ def build_notification_footer(self, recipient: Team | User) -> str:
...
+    notification: BaseNotification, **kwargs: Any
+) -> BaseNotification:
+    notification_copy = deepcopy(notification)
+
+    assert "url_format" in kwargs
+
+    for property in kwargs:
+        setattr(notification_copy, property, kwargs[property])
+
+    assert getattr(notification_copy, "url_format") == "<{url}|{text}>"
```
[the remark this answers is not in the document]
Comment (97varun) [reply src/sentry/notifications/notifications/base.py]: Removed.

**Pull request merged**